Kentor.AuthServices/SAML2P/Saml2Response.cs

```diff
@@ -1,480 +1,487 @@
+            }
+            else
+            {
+                // If the response message is not signed, all assersions have to be signed correctly
+                foreach (var assertionNode in AllAssertionElementNodes)
+                {
+                  CheckSignature(assertionNode, idpKeys);
+                }
+            }
+        }
+
+        private static readonly string[] allowedTransforms = new string[]
+        {
+            SignedXml.XmlDsigEnvelopedSignatureTransformUrl,
+            SignedXml.XmlDsigExcC14NTransformUrl,
+            SignedXml.XmlDsigExcC14NWithCommentsTransformUrl
+        };
+
+        /// <summary>Checks the signature.</summary>
+        /// <param name="signedRootElement">The signed root element.</param>
+        /// <param name="idpKeys">A list containing one ore more assymetric keys of a algorithm.</param>
+        private static void CheckSignature(XmlElement signedRootElement, IList<AsymmetricAlgorithm> idpKeys)
+        {
+            var xmlDocument = new XmlDocument { PreserveWhitespace = true };
+            xmlDocument.LoadXml(signedRootElement.OuterXml);
+
+            var signature = xmlDocument.DocumentElement["Signature", SignedXml.XmlDsigNamespaceUrl];
+            if (signature == null)
+            {
+                throw new Saml2ResponseFailedValidationException("The SAML Response is not signed and contains unsigned Assertions. Response cannot be trusted.");
+            }
+
+            var signedXml = new SignedXml(xmlDocument);
+            signedXml.LoadXml(signature);
+
+            var signedRootElementId = "#" + signedRootElement.GetAttribute("ID");
+
+            if (signedXml.SignedInfo.References.Count == 0)
+            {
+                throw new Saml2ResponseFailedValidationException("No reference found in Xml signature, it doesn't validate the Xml data.");
+            }
+
+            if (signedXml.SignedInfo.References.Count != 1)
+            {
+                throw new Saml2ResponseFailedValidationException("Multiple references for Xml signatures are not allowed.");
+            }
+
+            var reference = signedXml.SignedInfo.References.Cast<Reference>().Single();
+
+            if (reference.Uri != signedRootElementId)
+            {
+                throw new Saml2ResponseFailedValidationException("Incorrect reference on Xml signature. The reference must be to the root element of the element containing the signature.");
+            }
+
+            foreach (Transform transform in reference.TransformChain)
+            {
+                if (!allowedTransforms.Contains(transform.Algorithm))
+                {
+                    throw new Saml2ResponseFailedValidationException(
+                        "Transform \"" + transform.Algorithm + "\" found in Xml signature is not allowed in SAML.");
+                }
+            }
+
+            var validSignature = false;
+
+            for (var i = 0; !validSignature && i < idpKeys.Count; i++)

```

Comment:
or

    var validSignature = idpKeys.Any(key => signedXml.CheckSignature(key));

or even

    var validSignature = idpKeys.Any(signedXml.CheckSignature);

makes it more compact and readable